### ISSUE TITLE
[ 2 . 9 ]  tests skip

### ISSUE THREAD

**Issue by wyNomad** [flagos-ai/Torch-FL#190]
# torch 2.9 downgrade follow-up issues (draft)

> Generated from `fix_2.9_watch` @ `215d87e` (CI green [1/8]-[8/8]).
> The 11 `skipif(torch.__version__.startswith("2.9"))` skips are technical debt,
> NOT fixes. File these against `flagos-ai/PyTorch-Plugin-FL` (upstream) once the
> `fix_2.9_watch` fork branch is PRed, OR against `BrianPei/PyTorch-Plugin-FL`
> (fork) as internal tracking — decide per workflow.
> Two issues recommended (two independent root causes):
>   - Issue A: RNG dispatch incompatibility (10 tests, one root cause)
>   - Issue B: profiler key_averages() drops aten:: operator keys (1 test, independent)
> Before filing, search existing issues for duplicates (AI checklist requirement).

---

### Issue Type
- [x] Bug Report

### AI Agent Information
- **Agent**: Claude Code CLI
- **Model**: Claude Opus 5 (1M context)
- **Session Context**: torch 2.10 → 2.9 CUDA CI downgrade on `fix_2.9_watch`. After the
  three env fixes (LD_LIBRARY_PATH strip / `TORCH_DEVICE_BACKEND_AUTOLOAD=0` /
  `FLAGGEMS_KERNEL=0`), the build and import path are green, but 10 RNG tests fail under
  torch 2.9 because `torch_fl`'s RNG/generator injection was written for torch 2.10's
  dispatch semantics. They were conditionally skipped to get the line green and tracked here.

### Summary
Under torch 2.9, `torch_fl`'s flagos generator injection diverges from torch 2.10 in two
ways: (1) generator device-mismatch guards segfault instead of raising (3 tests), and
(2) `_like` RNG ops fall back to ATen's default generator and are not reproducible under
`torch.manual_seed` (7 tests). Same root cause, two surfaces. 10 tests are currently
`skipif(2.9)`-skipped on `fix_2.9_watch`; 2.10 baseline is unaffected.

### Environment
<details>
<summary>Expand environment details</summary>

- **Platform**: CUDA (boxing mode, vendor flag_gems native CUDA)
- **Python**: vendor interpreter (镜像自带)
- **PyTorch**: 2.9.0+cpu venv + external 2.9 ABI `libtorch_cuda.so` extracted from
  `torch==2.9.0+cu130` wheel (vendor 2.10 image is not rebuildable, no Dockerfile)
- **torch_fl**: `fix_2.9_watch` @ `215d87e`
- **Hardware**: NVIDIA CUDA runner (`nv-8g-cicd-pytorch`)

**Build config (key env):**
```bash
export FLAGGEMS_KERNEL=0        # 8baeee1: CUDA must take FLAGGEMS_PYTHON path under 2.9
export TORCH_DEVICE_BACKEND_AUTOLOAD=0  # e106dcd: probe stage, no device backend autoload
# 65cddd9: vendor(2.10) torch/lib stripped from LD_LIBRARY_PATH before first import torch
```
</details>

### Reproduction
```bash
# On fix_2.9_watch @ 215d87e, torch 2.9 env above:
pytest tests/integration/ops/test_rng_dispatch.py -m main_ops -v
```
Currently skipped (would fail without the skip):
```bash
# Surface 1 — segfault (skipif, a509c0f + a92e684):
pytest tests/integration/ops/test_rng_dispatch.py::TestRngSeedSource::test_flagos_generator_device_mismatch_raises
pytest tests/integration/ops/test_rng_dispatch.py::TestRngSeedSource::test_flagos_generator_factory_mismatch_raises
pytest tests/integration/ops/test_rng_dispatch.py::TestRngSeedSource::test_flagos_generator_does_not_reroute_cpu_op

# Surface 2 — reproducibility assertion failure (runtime skip in body, 623df9d):
pytest "tests/integration/ops/test_rng_dispatch.py::TestRngReproducible::test_same_seed_same_draw[rand_like]"
pytest "tests/integration/ops/test_rng_dispatch.py::TestRngReproducible::test_same_seed_same_draw[randn_like]"
pytest "tests/integration/ops/test_rng_dispatch.py::TestRngReproducible::test_same_seed_same_draw[randint_like]"
pytest "tests/integration/ops/test_rng_dispatch.py::TestRngReproducible::test_same_seed_same_draw[rand_like.out]"
pytest "tests/integration/ops/test_rng_dispatch.py::TestRngReproducible::test_same_seed_same_draw[randn_like.out]"
pytest "tests/integration/ops/test_rng_dispatch.py::TestRngReproducible::test_same_seed_same_draw[randint_like.out]"
pytest "tests/integration/ops/test_rng_dispatch.py::TestRngReproducible::test_same_seed_same_draw[randint_like.low_dtype_out]"
```

### Expected vs Actual Behavior
**Expected:** On torch 2.10 these 10 tests pass. On torch 2.9 they should also pass once
`torch_fl`'s RNG injection is adapted to 2.9's generator dispatch.
**Actual (torch 2.9):**
- Surface 1: `flagos` generator on a boxed-to-CUDA/CPU tensor infinite-redispatches to
  **SIGSEGV** (process-level signal, `xfail` cannot catch → must `skip`). torch 2.9 lacks
  the native generator device-mismatch check added in 2.10.
- Surface 2: `_draw(op, SEED) == _draw(op, SEED)` is False — `_like` RNG ops fall back to
  ATen's own default generator instead of the shared flagos generator, so they are not
  reproducible under `torch.manual_seed`. Non-`_like` `rand/randn/randint` pass.

### Root Cause Analysis
torch 2.9's RNG/generator dispatch mechanism differs systematically from 2.10.
`torch_fl`'s flagos generator injection logic was written for 2.10's dispatch semantics.
Surface 1 (mismatch guard segfault, 3 tests) and Surface 2 (`_like` fallback, 7 tests) are
two faces of the same root cause: the flagos generator is not wired into 2.9's dispatch path
the way 2.10 expects. Fixing one place should resolve both surfaces.

### Proposed Solution
Compare torch 2.9 vs 2.10 generator dispatch + `torch_fl`'s flagos generator injection code
(likely `csrc/aten/generated` and/or `torch_fl` python-side RNG injection), and adapt the
2.10-targeted injection to 2.9's dispatch mechanism. Specifically:
- Restore the generator device-mismatch guard behavior under 2.9 (so mismatched
  generator→tensor combinations raise `RuntimeError("device type for generator")` instead
  of infinite-redispatch→SIGSEGV).
- Route `_like` RNG ops through the shared flagos generator (seeded by
  `torch.manual_seed`) instead of ATen's default generator, so they become reproducible.
This is backend/operator-development work, not CI configuration. `setup.py` CUDA defaults
must NOT change (2.10 baseline protection); the `FLAGGEMS_KERNEL=0` env override stays as-is.

### Verification Plan
- [ ] Unit test: generator device-mismatch raises `RuntimeError` under torch 2.9 (no segfault).
- [ ] Integration test: `test_same_seed_same_draw[_like*]` passes under torch 2.9.
- [ ] Regression: the same 10 tests still pass under torch 2.10 (skipif removed, both lines run).
- [ ] Manual: `torch.manual_seed(42); a=torch.randn_like(x, device="flagos:0"); b=torch.randn_like(x, device="flagos:0"); assert torch.equal(a,b)`.

### Context & Investigation
- Branch `fix_2.9_watch`, base = working 2.10 line `eea6727`.
- Env fixes (separate, already green): `65cddd9` (LD_LIBRARY_PATH strip),
  `e106dcd` (autoload disable), `8baeee1` (FLAGGEMS_KERNEL=0). These are CI/env layer;
  this issue is the code-adaptation layer that env fixes cannot reach.
- Skips introduced by: `a509c0f` (device_mismatch + factory_mismatch), `a92e684`
  (does_not_reroute_cpu_op), `623df9d` (7 `_like`).
- Full handoff: `fix_2.9_summary.md` in the contributor's study notes.

### Related Code Locations
- `tests/integration/ops/test_rng_dispatch.py:215` — `_like` runtime skip (623df9d)
- `tests/integration/ops/test_rng_dispatch.py:305` / `:326` — mismatch skipif (a509c0f)
- `tests/integration/ops/test_rng_dispatch.py:345` — reroute skipif (a92e684)
- Fix target (investigate): `csrc/aten/generated/` (RNG registration) and `torch_fl/`
  python-side flagos generator injection.

### Checklist
- [x] Environment information provided
- [x] Reproducer included (pytest commands + env)
- [x] Error behavior described (segfault / assertion failure)
- [x] Root cause analyzed
- [x] Proposed solution with approach
- [x] Affected code locations with line numbers + commits
- [x] Verification plan described
- [ ] Checked for duplicate issues (do before filing)
- [x] All text in English

---

### Issue Type
- [x] Bug Report

### AI Agent Information
- **Agent**: Claude Code CLI
- **Model**: Claude Opus 5 (1M context)
- **Session Context**: Same torch 2.10 → 2.9 CUDA CI downgrade session. After env fixes,
  profiler/tracer itself works, but one profiler-parity test fails under torch 2.9 because
  `key_averages()` no longer surfaces `aten::` CPU operator keys. Skipped to keep the line
  green; tracked here. Independent root cause from Issue A.

### Summary
Under torch 2.9, `prof.key_averages()` aggregates only device kernel + cuda runtime events
(`ampere_sgemm_...`, `cudaLaunchKernel`, `cudaMemcpy`, ...) and emits **no `aten::`-prefixed
CPU operator key** (2.10 emits `aten::mm`). The `test_device_time_attribution` test
cross-checks `aten::mm`'s device time against linked device events; with no operator key to
anchor on, it cannot run. 1 test currently `skipif(2.9)`-skipped; the other 6 tests in the
same file pass (profiler/tracer itself is fine). 2.10 baseline unaffected.

### Environment
<details>
<summary>Expand environment details</summary>

Same as Issue A: CUDA / torch 2.9.0+cpu + external 2.9 `libtorch_cuda.so` / `fix_2.9_watch@215d87e`.
</details>

### Reproduction
```bash
pytest tests/integration/test_profiler_parity.py::test_device_time_attribution -v
```
The test asserts `prof.key_averages()` contains an `aten::mm` key; under 2.9 it does not.

### Expected vs Actual Behavior
**Expected:** `key_averages()` includes `aten::mm` (as on 2.10), so device-time attribution
can cross-check against the operator key. **Actual (torch 2.9):** `key_averages()` returns
only device kernel + cuda runtime events; no `aten::` operator key. The other 6 tests in
`test_profiler_parity.py` pass → the profiler/tracer works; this is purely a
`key_averages()` aggregation-behavior difference.

### Root Cause Analysis
torch 2.9's `key_averages()` aggregation does not include CPU `aten::` operator keys that
2.10 includes. Two open questions for whoever owns this (needs first triage):
1. Is the 2.9 aggregation granularity genuinely different (test assertion should adapt to
   the new 2.9 shape — leans QA/test-engineering)?
2. Or is `torch_fl`'s profiler integration (e.g. `record_function` under boxing dispatch)
   not firing the `aten::` key under 2.9 (leans backend/dev)?

This is a gray-area item: classify after first triage. It is independent of Issue A's RNG
root cause.

### Proposed Solution
First, determine whether 2.9 legitimately changed `key_averages()` aggregation (compare 2.9
vs 2.10 on a plain CUDA torch, no torch_fl, with an `aten::mm` workload). If 2.9 plain torch
also drops the `aten::` key → adapt the test assertion to 2.9's new aggregation shape
(anchor on a device kernel or runtime event instead of `aten::mm`). If 2.9 plain torch still
has `aten::mm` but torch_fl-2.9 does not → investigate `torch_fl`'s profiler integration
under 2.9 boxing dispatch.

### Verification Plan
- [ ] Reproduce on plain torch 2.9 CUDA (no torch_fl): does `key_averages()` include `aten::mm`?
- [ ] Integration test: `test_device_time_attribution` passes under torch 2.9 after fix.
- [ ] Regression: same test passes under torch 2.10 (skipif removed).

### Context & Investigation
- Skip introduced by `215d87e` (HEAD of `fix_2.9_watch`).
- 6 sibling tests in `test_profiler_parity.py` pass under 2.9 → profiler/tracer healthy.
- Full handoff: `fix_2.9_summary.md` §IV.B / §V.2 in the contributor's study notes.

### Related Code Locations
- `tests/integration/test_profiler_parity.py:298` — skipif (215d87e)
- `tests/integration/test_profiler_parity.py:308` — `test_device_time_attribution`
- Fix target (investigate): `torch_fl` profiler integration path under 2.9, and 2.9 vs 2.10
  `key_averages()` aggregation semantics.

### Checklist
- [x] Environment information provided
- [x] Reproducer included
- [x] Error behavior described
- [x] Root cause analyzed (with explicit open question for triage)
- [x] Proposed solution with approach
- [x] Affected code locations with line numbers + commit
- [x] Verification plan described
- [ ] Checked for duplicate issues (do before filing)
- [x] All text in English
